python/cog/server/worker.py

```diff
@@ -111,43 +229,52 @@ def _assert_state(self, state: WorkerState) -> None:
                 f"Invalid operation: state is {self._state} (must be {state})"
             )
 
-    def _wait(
-        self, poll: Optional[float] = None, raise_on_error: Optional[str] = None
-    ) -> Iterator[_PublicEventType]:
-        done = None
-
-        if poll:
-            send_heartbeats = True
-        else:
-            poll = 0.1
-            send_heartbeats = False
-
-        while self._child.is_alive() and not done:
-            if not self._events.poll(poll):
-                if send_heartbeats:
-                    yield Heartbeat()
-                continue
-            # this needs aioprocessing.Pipe or similar
-            # multiprocessing.Pipe is not async
-            ev = self._events.recv()
-            yield ev
-
-            if isinstance(ev, Done):
-                done = ev
-
-        if done:
-            if done.error and raise_on_error:
-                raise FatalWorkerException(raise_on_error + ": " + done.error_detail)
-            self._state = WorkerState.READY
-            self._allow_cancel = False
-
+    _read_events_task: "Optional[asyncio.Task[None]]" = None
+
+    def _ensure_event_reader(self) -> None:
+        def handle_error(task: "asyncio.Task[None]") -> None:
+            exc = task.exception()
+            if not exc:
+                return
+            # Re-raise the exception in order to more easily capture exc_info
+            try:
+                raise exc
+            except Exception:
+                logging.error("caught exception", exc_info=True)
+
+        if not self._read_events_task:
+            trace("started read events task")
+            self._read_events_task = asyncio.create_task(self._read_events())
+            self._read_events_task.add_done_callback(handle_error)
+            # and idk error handling or whatever
+
+    async def _read_events(self) -> None:
+        # in 3.10 Event started doing get_running_loop
+        # previously it stored the loop when created, which causes an error in tests
+        if sys.version_info.minor < 10:
+            self._terminating = asyncio.Event()
+            self._mux.shutdown = self._terminating
+
+        while self._child.is_alive() and not self._terminating.is_set():
+            try:
+                # this can still be running when the task is destroyed
+                id, event = await self._events.coro_recv_with_exit(self._terminating)
+                trace("recv", event)
+            except asyncio.CancelledError:
+                return
+            if id == "LOG" and "SETUP" in self._mux.outs:
+                id = "SETUP"
+            if id == "LOG" and len(self._predictions_in_flight) == 1:
+                id = list(self._predictions_in_flight)[0]
```

Comment:
some logs can have a .source if there's a suitable contextvar, but there are also logs for the entire batch that need to be discarded to avoid leaking information about other users' predictions